pkg/keystone/cronjob.go

```diff
@@ -0,0 +1,125 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keystone
+
+import (
+	keystonev1beta1 "github.com/openstack-k8s-operators/keystone-operator/api/v1beta1"
+	common "github.com/openstack-k8s-operators/lib-common/modules/common"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/affinity"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/env"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// TrustFlushCommand -
+	TrustFlushCommand = "/usr/local/bin/kolla_set_configs && keystone-manage trust_flush"
+)
+
+// CronJob func
+func CronJob(
+	instance *keystonev1beta1.KeystoneAPI,
+	labels map[string]string,
+	annotations map[string]string,
+) *batchv1.CronJob {
+	runAsUser := int64(0)
+
+	args := []string{"-c"}
+	if instance.Spec.Debug.Service {
+		args = append(args, common.DebugCommand)
+	} else {
+		args = append(args, TrustFlushCommand)
+	}
+
+	envVars := map[string]env.Setter{}
+	envVars["KOLLA_CONFIG_FILE"] = env.SetValue(KollaConfig)
+	envVars["KOLLA_CONFIG_STRATEGY"] = env.SetValue("COPY_ALWAYS")
+
+	parallelism := int32(1)
+	completions := int32(1)
+
+	cronjob := &batchv1.CronJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ServiceName + "-cron",
+			Namespace: instance.Namespace,
+		},
+		Spec: batchv1.CronJobSpec{
+			Schedule:          "1 * * * *",
+			ConcurrencyPolicy: batchv1.ForbidConcurrent,
+			JobTemplate: batchv1.JobTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: annotations,
+					Labels:      labels,
+				},
+				Spec: batchv1.JobSpec{
+					Parallelism: &parallelism,
+					Completions: &completions,
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Name:  ServiceName + "-cron",
+									Image: instance.Spec.ContainerImage,
+									Command: []string{
+										"/bin/bash",
+									},
+									Args:         args,
+									Env:          env.MergeEnvs([]corev1.EnvVar{}, envVars),
+									VolumeMounts: getVolumeMounts(),
+									SecurityContext: &corev1.SecurityContext{
+										RunAsUser: &runAsUser,
+									},
+								},
+							},
+							Volumes:            getVolumes(instance.Name),
+							RestartPolicy:      corev1.RestartPolicyNever,
+							ServiceAccountName: ServiceAccount,
+						},
+					},
+				},
+			},
+		},
+	}
+	// If possible two pods of the same service should not
+	// run on the same worker node. If this is not possible
+	// the get still created on the same worker node.
+	cronjob.Spec.JobTemplate.Spec.Template.Spec.Affinity = affinity.DistributePods(
+		common.AppSelector,
+		[]string{
+			ServiceName + "-cron",
+		},
+		corev1.LabelHostname,
+	)
```

Comment:
is affinity required if the Parallelism is 1?

Reply:
No. I'll remove this.

Just for record.
We generally do not require multiple pods to run cron jobs. The only exception would be one used by glance which should be run at the same node(with /var/lib/glance) shared with glance-api but that don't require affinity either.